### PR TITLE
feat: Add chunk endorsements bitmap to BlockHeader

### PR DESCRIPTION
**Pull request by tayfunelmas**
Previously we added `ChunkEndorsementBitmap` to `BlockInfo` to pass the information about which endorsements are received to the `EpochManager`. We used this in a tool to simulate the validator rewards and kickoffs if we used the received endorsements stats (instead of chunk production stats).

In this PR, we complete the picture by adding the bitmap to the `BlockHeader`. This bitmap is passed to the `BlockInfo` when saving it for the `EpochManager`. The bitmap is produced directly from the endorsement signatures in the `BlockBody`, with the same number of shards and chunk validator assignments per shard.

We also add validation logic for block body and header. For the block body, we can compare the endorsement signatures to the bitmap values. For the block header, we can compare the length of the bitmap to the number of validator assignments.

NOTE: I am currently reusing `ProtocolFeature::ChunkEndorsementsInBlockHeader` for the new block header version. And due to some tests assuming that only the last block header version is used, I moved this feature beyond Nightly. Once I can verify the easy mode works, I will send a PR to introduce a separate feature to start using `BlockHeaderV5` and in Nightly.

### CODE REVIEW

File: chain/chain/src/stateless_validation/chunk_endorsement.rs [hunk 1]
```diff
@@ -101,6 +112,60 @@ pub fn validate_chunk_endorsements_in_block(
...
+        return Err(Error::InvalidChunkEndorsementBitmap(
+            format!("Number of shards in bitmap and in epoch do not match: shards in bitmap={}, shards in epoch={}",
+                chunk_endorsements.num_shards(), shard_ids.len())));
+    }
+    for shard_id in shard_ids.into_iter() {
+        let num_validator_assignments = epoch_manager
+            .get_chunk_validator_assignments(&epoch_id, shard_id, header.height())?
+            .len();
+        // 1. Bitmap's length must be equal or greater than the number of assignments in the previous block.
+        if chunk_endorsements.len(shard_id).unwrap() < num_validator_assignments {
```
Comment (Longarithm) [inline chain/chain/src/stateless_validation/chunk_endorsement.rs:156]: This currently allows very long bitmap with `false` tail. We need precise length check (`ceil(num_assignments / 8) * 8`?) and unit test for that.

File: core/primitives/src/block_header.rs [hunk 2]
```diff
@@ -493,6 +591,7 @@ impl BlockHeader {
         block_merkle_root: CryptoHash,
         prev_height: BlockHeight,
         clock: near_time::Clock,
+        chunk_endorsements: Option<ChunkEndorsementsBitmap>,
```
Comment (Longarithm) [inline core/primitives/src/block_header.rs:594]: This function and its `genesis` version became really big... Could you leave one or two latest versions of block headers on top and hide old ones into something like `old_impl`?

**Review by Longarithm — APPROVED** [review]
Approving to unblock but please look into the comment about long `false` tail.
Also I think it's a good time to fix the "block header version" issue, we started to discuss it offline.

File: core/primitives/src/block_header.rs (hunk 2 above)
Comment (tayfunelmas) [reply core/primitives/src/block_header.rs:594]: done.

File: chain/chain/src/stateless_validation/chunk_endorsement.rs (hunk 1 above)
Comment (tayfunelmas) [reply chain/chain/src/stateless_validation/chunk_endorsement.rs:156]: good point, changing the check and adding tests.

Comment (tayfunelmas) [reply chain/chain/src/stateless_validation/chunk_endorsement.rs:156]: actually this requires being able to modify a BlockHeaderV5, so I will do that in follow-up PR (in which I will add mutability to different header versions). stay tuned.

**Pull request merged**